src/app/Mappings/components/AddEditMappingModal.tsx

```diff
@@ -1,36 +1,58 @@
 import * as React from 'react';
-import {
-  Modal,
-  Button,
-  Form,
-  FormGroup,
-  TextInput,
-  Grid,
-  GridItem,
-  Alert,
-} from '@patternfly/react-core';
+import * as yup from 'yup';
+import { Modal, Button, Form, FormGroup, Grid, GridItem, Alert } from '@patternfly/react-core';
 import spacing from '@patternfly/react-styles/css/utilities/Spacing/spacing';
+
 import SimpleSelect, { OptionWithValue } from '@app/common/components/SimpleSelect';
 import { MappingBuilder, IMappingBuilderItem } from './MappingBuilder';
 import { getMappingFromBuilderItems } from './MappingBuilder/helpers';
-import { MappingType, IOpenShiftProvider, IVMwareProvider } from '@app/queries/types';
+import { Mapping, MappingType, IOpenShiftProvider, IVMwareProvider } from '@app/queries/types';
 import { useProvidersQuery, useMappingResourceQueries } from '@app/queries';
 import { updateMockStorage } from '@app/queries/mocks/helpers';
 import './AddEditMappingModal.css';
 import { usePausedPollingEffect } from '@app/common/context';
 import LoadingEmptyState from '@app/common/components/LoadingEmptyState';
+import { useFormField, useFormState } from '@app/common/hooks/useFormState';
+import ValidatedTextInput from '@app/common/components/ValidatedTextInput';
 
 interface IAddEditMappingModalProps {
   title: string;
   onClose: () => void;
   mappingType: MappingType;
 }
 
+const useMappingFormState = () => ({
+  mapping: useFormState({
+    name: useFormField('', yup.string().label('Mapping name').required()),
+    sourceProvider: useFormField<IVMwareProvider | null>(
+      null,
+      yup.mixed<IVMwareProvider>().label('Source provider').required()
+    ),
+    targetProvider: useFormField<IOpenShiftProvider | null>(
+      null,
+      yup.mixed<IOpenShiftProvider>().label('Target provider').required()
+    ),
+  }),
+  networkMapping: useFormState({
+    mapping: useFormField<Mapping | null>(null, yup.mixed<Mapping>().required()),
+    isSaveNewMapping: useFormField(false, yup.boolean().required()),
+    newMappingName: useFormField('', yup.string()),
+  }),
+  storageMapping: useFormState({
+    mapping: useFormField<Mapping | null>(null, yup.mixed<Mapping>().required()),
+    isSaveNewMapping: useFormField(false, yup.boolean().required()),
+    newMappingName: useFormField('', yup.string()),
+  }),
+});
+
+export type MappingFormState = ReturnType<typeof useMappingFormState>;
```

Comment:
We don't necessarily need to export this type or even have this form lifted into a `useMappingFormState`, that was just useful to have `PlanWizardFormState` as a prop type for passing forms down into the wizard steps.

Also I think we only need one `useFormState` in this component, we can remove the `networkMapping` and `storageMapping` useFormState calls here because the mapping field will be part of that singular form (the one you call `mapping` right now), and its fields are redundant (isSaveNewMapping is only relevant in the wizard, `newMappingName` is redundant with `name`, and we don't need separate state for network vs storage since the modal will just be using one at a time).

So I think the whole thing can be replaced with:

```ts
const form = useFormState({
  name: ...
  sourceProvider,
  targetProvider,
  // and then in my PR i'll add `mapping` here
});
```

and it can probably go right inside the component below instead of having `useMappingFormState`, just like you did in SelectNetworkModal (although that doesn't really matter either way)

Reply:
Ok, actually it's interesting how we can use file level variables and export them to lift up state.

Yeah you're right this ended up over-complicated because I started addressing MappingBuilder which is not happening in this PR anymore.